File: pulsar-common/src/main/java/ai/platon/pulsar/common/config/VolatileConfig.java
package ai.platon.pulsar.common.config;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Created by vincent on 18-1-17.
 * Copyright @ 2013-2017 Platon AI. All rights reserved
 */
public class VolatileConfig extends MutableConfig {

    public static final VolatileConfig EMPTY = new VolatileConfig();

    private ImmutableConfig fallbackConfig;
    private Map<String, Integer> ttls = Collections.synchronizedMap(new HashMap<>());
    private Map<String, Object> variables = new HashMap<>();

    public VolatileConfig() {
        super(false);
    }

    public VolatileConfig(ImmutableConfig fallbackConfig) {
        super(false);
        this.fallbackConfig = fallbackConfig;
    }

    public void reset() {
        ttls.clear();
        variables.clear();
        super.clear();
    }

    @Nonnull
    @Override
    public String get(String name, String defaultValue) {
        Objects.requireNonNull(name);
        Objects.requireNonNull(defaultValue);

        String value = super.get(name);
        if (value != null) {
            if (!isExpired(name)) {
                return super.get(name, defaultValue);
            } else {
                if (LOG.isTraceEnabled()) {
                    LOG.trace("Session config (with default) {} is expired", name);
                }
                ttls.remove(name);
                super.unset(name);
            }
        }

        return fallbackConfig != null ? fallbackConfig.get(name, defaultValue) : defaultValue;
    }

    @Nullable
    @Override
    public String get(String name) {
        Objects.requireNonNull(name);

        String value = super.get(name);
        if (value != null) {
            if (!isExpired(name)) {
                return value;
            } else {
                if (LOG.isTraceEnabled()) {
                    LOG.trace("Session config {} is expired", name);
                }
                ttls.remove(name);
                super.unset(name);
            }
        }

        return fallbackConfig != null ? fallbackConfig.get(name) : null;
    }

    public void set(String name, String value, int ttl) {
        Objects.requireNonNull(name);
        Objects.requireNonNull(value);

        setTTL(name, ttl);
        super.set(name, value);
    }

    @Nullable
    public String getAndSet(String name, String value, int ttl) {
        Objects.requireNonNull(name);
        Objects.requireNonNull(value);

        String old = get(name);
        if (old != null) {
            this.set(name, value, ttl);
        }
        return old;
    }

    public int getTTL(String name) {
        return ttls.getOrDefault(name, Integer.MAX_VALUE);
    }

    public void setTTL(String name, int ttl) {
        if (ttl > 0) {
            ttls.put(name, ttl);
        } else {
            ttls.remove(name);
            super.unset(name);
        }
    }

    public Object getVariable(String name) {
        return variables.get(name);
    }

    public <T> Object putBean(T bean) {
        return putBean(bean.getClass().getName(), bean);
    }

    public <T> Object putBean(String name, T bean) {
        return variables.put(name, bean);
    }

    @Nullable
    public <T> T getBean(Class<T> bean) {
        return getBean(bean.getName(), bean);
    }

    @Nullable
    public <T> T getBean(String name, Class<T> bean) {
        Object obj = variables.get(name);
        if (obj != null && bean.isAssignableFrom(obj.getClass())) {
            return (T)obj;
        }
        return null;
    }

    public void setVariable(String name, Object value) {
        variables.put(name, value);
    }

    public boolean isExpired(String key) {
        return false;
    }

    @Nullable
    public ImmutableConfig getFallbackConfig() {
        return fallbackConfig;
    }

    public void setFallbackConfig(ImmutableConfig fallbackConfig) {
        this.fallbackConfig = fallbackConfig;
    }
}

File: pulsar-common/src/main/kotlin/ai/platon/pulsar/common/config/VolatileConfig.kt
package ai.platon.pulsar.common.config

import java.util.concurrent.ConcurrentHashMap
import kotlin.reflect.KClass

/**
 * Created by vincent on 18-1-17.
 * Copyright @ 2013-2017 Platon AI. All rights reserved
 */
open class VolatileConfig : MutableConfig {
    var fallbackConfig: ImmutableConfig? = null

    private val ttls: MutableMap<String, Int> = ConcurrentHashMap()
    val variables: MutableMap<String, Any> = ConcurrentHashMap()

    constructor(): this("", false, listOf())

    constructor(loadDefaults: Boolean): this(
        System.getProperty(CapabilityTypes.LEGACY_CONFIG_PROFILE, ""),
        loadDefaults
    )

    constructor(
        profile: String = System.getProperty(CapabilityTypes.LEGACY_CONFIG_PROFILE, ""),
        loadDefaults: Boolean = true,
        resources: Iterable<String> = DEFAULT_RESOURCES
    ): super(profile, loadDefaults, resources)

    constructor(fallbackConfig: ImmutableConfig) : this("", false) {
        this.fallbackConfig = fallbackConfig
        if (fallbackConfig is VolatileConfig) {
            ttls.putAll(fallbackConfig.ttls)
            variables.putAll(fallbackConfig.variables)
        }
    }

    fun reset() {
        ttls.clear()
        variables.clear()
        super.clear()
    }

    override fun get(name: String, defaultValue: String): String {
        val value = super.get(name)
        if (value != null) {
            if (!isExpired(name)) {
                return super.get(name, defaultValue)
            } else {
                if (logger.isTraceEnabled) {
                    logger.trace("Session config (with default) {} is expired", name)
                }
                ttls.remove(name)
                super.unset(name)
            }
        }

        return fallbackConfig?.get(name, defaultValue) ?: defaultValue
    }

    override fun get(name: String): String? {
        val value = super.get(name)
        if (value != null) {
            if (!isExpired(name)) {
                return value
            } else {
                if (logger.isTraceEnabled) {
                    logger.trace("Session config {} is expired", name)
                }
                ttls.remove(name)
                super.unset(name)
            }
        }
        return fallbackConfig?.get(name)
    }

    operator fun set(name: String, value: String, ttl: Int) {
        setTTL(name, ttl)
        super.set(name, value)
    }

    fun getAndSet(name: String, value: String, ttl: Int): String? {
        val old = get(name)
        if (old != null) {
            this[name, value] = ttl
        }
        return old
    }

    fun getTTL(name: String): Int {
        return ttls.getOrDefault(name, Int.MAX_VALUE)
    }

    open fun setTTL(name: String, ttl: Int) {
        if (ttl > 0) {
            ttls[name] = ttl
        } else {
            ttls.remove(name)
            super.unset(name)
        }
    }

    fun <T : Any> putBean(bean: T): Any? {
        return putBean(bean.javaClass.name, bean)
    }

    fun <T : Any> putBean(name: String, bean: T): Any? {
        return variables.put(name, bean)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getBeanOrNull(bean: Class<T>): T? {
        val obj = variables.values.firstOrNull { bean.isAssignableFrom(it.javaClass) }
        return obj as? T
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : Any> getBeanOrNull(bean: KClass<T>): T? {
        val obj = variables.values.firstOrNull { bean.java.isAssignableFrom(it.javaClass) }
        return obj as? T
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getBeanOrNull(name: String, bean: Class<T>): T? {
        val obj = variables[name]
        return if (obj != null && bean.isAssignableFrom(obj.javaClass)) {
            obj as T
        } else null
    }

    fun <T : Any> getBeanOrNull(name: String, bean: KClass<T>): T? {
        return getBeanOrNull(name, bean.java)
    }

    fun <T : Any> removeBean(bean: T): Any? {
        return variables.remove(bean.javaClass.name)
    }

    fun <T : Any> removeBean(name: String): Any? {
        return variables.remove(name)
    }

    fun getVariable(name: String?): Any? {
        return variables[name]
    }

    fun setVariable(name: String, value: Any) {
        variables[name] = value
    }

    open fun isExpired(key: String): Boolean {
        return false
    }

    companion object {
        val EMPTY = VolatileConfig()
        val UNSAFE = VolatileConfig()
    }
}
